File: web/js/replica.js
import { app } from "../../../scripts/app.js";
import { api } from "../../../scripts/api.js";
import { analysisQueue, draftFor, labels, newId, normalizeLanguage } from "./replica_model.js";
import { fileUrl, invalidateReplicaEditor, openReplicaEditor, refreshReplicaEditorLocale, replicaRequest } from "./replica_editor.js";

const REVERSE = "MengBaoImageReverse", SETTINGS = "MengBaoImageReplicaSettings", AUDIT = "MengBaoReplicaAudit";
const CLASSES = new Set([REVERSE, SETTINGS, AUDIT]);
let capabilities = {};
function language() { return normalizeLanguage(app.ui?.settings?.getSettingValue?.("Comfy.Locale")); }
function nodeClass(node) { return node?.comfyClass || node?.type; }
function widget(node, name) { return node.widgets?.find(value => value.name === name); }
function graphNode(id) { return app.graph?.getNodeById?.(id); }
function linkById(id) { return app.graph?.links?.get?.(id) || app.graph?.links?.[id]; }
function linkedAnalysis(node) {
  const input = node.inputs?.find(input => input.name === "analysis");
  return graphNode(linkById(input?.link)?.origin_id);
}
function currentDraft(node) {
  try { return JSON.parse(widget(node, "draft_json")?.value || "null"); }
  catch { return null; }
}
function storeDraft(node, draft) {
  const control = widget(node, "draft_json");
  if (control) control.value = JSON.stringify(draft);
  app.graph?.setDirtyCanvas?.(true, true);
  renderNode(node);
}
function modelLimit(node) {
  const output = node.outputs?.find(output => output.name === "replica_settings");
  const limits = (output?.links || []).map(id => graphNode(linkById(id)?.target_id)).filter(node => nodeClass(node) === "WANGImageAPI")
    .map(node => capabilities[widget(node, "model_type")?.value]).filter(Number.isFinite);
  return limits.length ? Math.min(...limits) : Math.max(16, ...Object.values(capabilities));
}
function hideTechnicalWidgets(node) {
  for (const name of ["request_id", "draft_json", "draft_id"]) {
    const control = widget(node, name);
    if (!control) continue;
    control.type = "hidden"; control.hidden = true;
    control.computeSize = () => [0, -4]; control.draw = () => {};
    if (control.inputEl) control.inputEl.style.display = "none";
    if (control.element) control.element.style.display = "none";
  }
}
function renderNode(node) {
  const state = node._mengBaoReplicaState, controls = node._mengBaoReplicaControls;
  if (!state || !controls) return;
  const text = labels(language()), type = nodeClass(node);
  state.language = language();
  node.title = type === REVERSE ? text.reverseTitle : type === SETTINGS ? text.settingsTitle : text.auditTitle;
  const names = { image: text.source, analysis: text.analysis, images: text.images, replica_settings: text.settings };
  for (const input of node.inputs || []) if (names[input.name]) input.label = input.localized_name = names[input.name];
  const outputs = type === REVERSE ? [text.analysis, text.reversePrompt] : type === SETTINGS ? [text.prompt, text.settings] : [text.images, text.report];
  for (const [index, output] of (node.outputs || []).entries()) output.label = output.localized_name = outputs[index];
  if (type === AUDIT) {
    controls.report.textContent = state.report || text.empty;
    controls.status.textContent = state.error || (state.stage ? `${text[state.stage] || text.requesting} · ${state.model || ""}` : state.report ? text.auditReady : "");
  } else {
    const record = state.record;
    controls.image.hidden = !record; controls.empty.hidden = Boolean(record); controls.empty.textContent = text.empty;
    if (record) { const url = fileUrl(record.source.id); if (controls.image.getAttribute("src") !== url) controls.image.src = url; }
    controls.button.textContent = type === SETTINGS ? text.open : record ? text.reanalyze : text.analyze;
    controls.button.disabled = state.busy || (type === SETTINGS && !record);
    const ready = type === SETTINGS ? currentDraft(node)?.confirmed ? text.confirmed : text.unconfirmed : text.ready;
    const attempts = (record?.attempts || []).filter(attempt => !attempt.success).map(attempt => `\n${attempt.model}: ${attempt.error}`).join("");
    controls.status.textContent = state.error || (state.busy ? `${text[state.stage] || text.requesting} · ${state.model || "gem-3.7-flash"}${state.previousError ? `\n${text.switching}\n${state.previousError}` : ""}` : record ? `${ready} · ${record.model}\n${record.analysis.elements.length} · ${text.analysis}${attempts}` : "");
  }
  node.setDirtyCanvas?.(true, true);
}
function refreshSettings(node) {
  const source = linkedAnalysis(node), record = source?._mengBaoReplicaState?.record;
  if (record && widget(source, "request_id")?.value === record.id) {
    const old = node._mengBaoReplicaState.record;
    node._mengBaoReplicaState.record = record;
    const draft = draftFor(record, currentDraft(node));
    if (old?.id && old.id !== record.id) invalidateReplicaEditor(old.id);
    storeDraft(node, draft);
  } else { node._mengBaoReplicaState.record = null; renderNode(node); }
}
function receiveAnalysis(node, record) {
  if (!record?.analysis || record.id !== widget(node, "request_id")?.value) return;
  Object.assign(node._mengBaoReplicaState, { record, busy: false, stage: "", error: "", model: record.model, previousError: "" });
  renderNode(node);
  for (const target of app.graph?._nodes || []) if (nodeClass(target) === SETTINGS && linkedAnalysis(target) === node) refreshSettings(target);
}
async function analyze(node) {
  const state = node._mengBaoReplicaState;
  if (state.busy) return;
  widget(node, "request_id").value = newId();
  Object.assign(state, { busy: true, stage: "requesting", model: "gem-3.7-flash", error: "", previousError: "" });
  renderNode(node);
  for (const target of app.graph?._nodes || []) if (nodeClass(target) === SETTINGS && linkedAnalysis(target) === node) refreshSettings(target);
  try {
    const prompt = await app.graphToPrompt();
    const payload = { ...analysisQueue(node.id, prompt.output), client_id: api.clientId };
    const response = await api.fetchApi("/prompt", { method: "POST", headers: { "Content-Type": "application/json" }, body: JSON.stringify(payload) });
    const data = await response.json();
    if (!response.ok || data.error) throw new Error(JSON.stringify(data.error || data.node_errors || `HTTP ${response.status}`));
    state.promptId = data.prompt_id;
  } catch (error) { state.busy = false; state.error = error.message; renderNode(node); }
}
async function edit(node) {
  refreshSettings(node);
  const state = node._mengBaoReplicaState;
  if (!state.record) { state.error = labels(language()).noAnalysis; renderNode(node); return; }
  try {
    capabilities = await replicaRequest("/capabilities");
    const record = state.record;
    await openReplicaEditor({ record, draft: draftFor(record, currentDraft(node)), draftId: widget(node, "draft_id").value,
      onChange: draft => storeDraft(node, draft), language, limit: () => modelLimit(node) });
  } catch (error) { state.error = error.message; renderNode(node); }
}
async function restore(node) {
  try {
    if (nodeClass(node) === REVERSE && widget(node, "request_id")?.value) {
      const id = widget(node, "request_id").value;
      const record = await replicaRequest(`/analysis/${id}`);
      if (widget(node, "request_id")?.value === id) receiveAnalysis(node, record);
    } else if (nodeClass(node) === SETTINGS) {
      if (!currentDraft(node)) { const draft = await replicaRequest(`/drafts/${widget(node, "draft_id").value}`); if (draft.analysis_id) storeDraft(node, draft); }
      refreshSettings(node);
    }
  } catch (error) { node._mengBaoReplicaState.error = error.message; renderNode(node); }
}
function configure(node) {
  hideTechnicalWidgets(node);
  if (nodeClass(node) === SETTINGS && !widget(node, "draft_id")?.value) widget(node, "draft_id").value = newId();
  if (!node._mengBaoReplicaControls) {
    node._mengBaoReplicaState = { record: null, busy: false, error: "", report: "" };
    const root = document.createElement("div"); root.className = "mengbao-replica-node";
    const status = document.createElement("div"); status.className = "replica-node-status"; status.setAttribute("role", "status");
    const controls = { root, status };
    if (nodeClass(node) === AUDIT) { controls.report = document.createElement("pre"); controls.report.className = "replica-node-report"; root.append(controls.report); }
    else {
      controls.image = document.createElement("img"); controls.image.alt = "Source"; controls.image.hidden = true;
      controls.empty = document.createElement("div"); controls.empty.className = "replica-node-empty";
      controls.button = document.createElement("button"); controls.button.type = "button";
      controls.button.addEventListener("click", () => nodeClass(node) === REVERSE ? analyze(node) : edit(node));
      root.append(controls.image, controls.empty, controls.button);
    }
    root.append(status); node._mengBaoReplicaControls = controls;
    node.addDOMWidget("mengbao_replica_controls", "mengbao-replica", root, { serialize: false, hideOnZoom: false, getMinHeight: () => 350, getMaxHeight: () => 350 });
    node.setSize?.([Math.max(360, node.size?.[0] || 0), node.computeSize?.()[1] || 410]);
  }
  renderNode(node); void restore(node);
}
function installListeners() {
  api.addEventListener?.("mengbao_replica_status", ({ detail }) => {
    const node = graphNode(detail.node_id), state = node?._mengBaoReplicaState;
    if (!state || (detail.request_id && detail.request_id !== widget(node, "request_id")?.value)) return;
    Object.assign(state, { stage: detail.stage, model: detail.model, previousError: detail.previous_error || "" }); renderNode(node);
  });
  api.addEventListener?.("executed", ({ detail }) => {
    const node = graphNode(detail.node);
    if (detail.output?.mengbao_replica_analysis && node?._mengBaoReplicaState) receiveAnalysis(node, detail.output.mengbao_replica_analysis[0]);
    if (detail.output?.mengbao_replica_report && node?._mengBaoReplicaState) { Object.assign(node._mengBaoReplicaState, { report: detail.output.mengbao_replica_report[0], stage: "", error: "" }); renderNode(node); }
    const notice = detail.output?.mengbao_replica_ratio?.[0];
    if (notice) {
      const [w, h] = notice.ratio.split(":").map(Number);
      if (notice.auto_resolution_mapped || Math.abs(w / h - notice.source_width / notice.source_height) > 1e-9) app.extensionManager?.toast?.add?.({ severity: "info", summary: labels(language()).ratio, detail: `${notice.source_width}×${notice.source_height} → ${notice.ratio}${notice.auto_resolution_mapped ? ` · ${labels(language()).autoResolution}` : ""}`, life: 9000 });
    }
  });
  for (const eventName of ["execution_error", "execution_interrupted"]) api.addEventListener?.(eventName, ({ detail }) => {
    for (const node of app.graph?._nodes || []) {
      const state = node._mengBaoReplicaState;
      if (!state || state.promptId !== detail.prompt_id) continue;
      state.busy = false; state.stage = ""; state.error = detail.exception_message || "Vision request cancelled"; renderNode(node);
    }
  });
  const locale = app.ui?.settings?.settingsLookup?.["Comfy.Locale"];
  if (locale) { const original = locale.onChange; locale.onChange = function (...args) { const result = original?.apply(this, args); setTimeout(() => { for (const node of app.graph?._nodes || []) if (CLASSES.has(nodeClass(node))) renderNode(node); refreshReplicaEditorLocale(); }, 0); return result; }; }
}
app.registerExtension({
  name: "MengBaoAI.replica",
  setup() {
    if (!document.querySelector("#mengbao-replica-style")) { const link = document.createElement("link"); link.id = "mengbao-replica-style"; link.rel = "stylesheet"; link.href = new URL("./replica.css", import.meta.url).href; document.head.append(link); }
    installListeners();
  },
  beforeRegisterNodeDef(nodeType, nodeData) {
    if (!CLASSES.has(nodeData.name)) return;
    for (const name of ["onNodeCreated", "onConfigure"]) {
      const original = nodeType.prototype[name]; nodeType.prototype[name] = function () { const result = original?.apply(this, arguments); configure(this); return result; };
    }
    const executed = nodeType.prototype.onExecuted;
    nodeType.prototype.onExecuted = function (message) {
      const result = executed?.apply(this, arguments);
      if (message?.mengbao_replica_analysis) receiveAnalysis(this, message.mengbao_replica_analysis[0]);
      if (message?.mengbao_replica_report) { Object.assign(this._mengBaoReplicaState, { report: message.mengbao_replica_report[0], stage: "" }); renderNode(this); }
      return result;
    };
    const connections = nodeType.prototype.onConnectionsChange;
    nodeType.prototype.onConnectionsChange = function () { const result = connections?.apply(this, arguments); if (nodeClass(this) === SETTINGS) setTimeout(() => refreshSettings(this), 0); return result; };
    const foreground = nodeType.prototype.onDrawForeground;
    nodeType.prototype.onDrawForeground = function () { if (this._mengBaoReplicaState?.language !== language()) renderNode(this); return foreground?.apply(this, arguments); };
  },
});
export { analyze, configure, modelLimit, receiveAnalysis, renderNode };

File: nodes/ecommerce/replica.py
import json

import torch

from ...api.vision_client import run_vision
from ...utils.image import image_tensor_to_png_bytes
from ...utils.replica import ANALYSIS_PROMPT, PROMPT_VERSION, audit_prompt, compile_settings, parse_analysis, parse_audit
from ...utils.replica_store import REPLICA_STORE, identity


def interrupted():
    try:
        from comfy.model_management import processing_interrupted
        return processing_interrupted()
    except ImportError:
        return False


def status(node_id, detail):
    if node_id is None:
        return
    try:
        from server import PromptServer
        if getattr(PromptServer, "instance", None):
            PromptServer.instance.send_sync("mengbao_replica_status", {"node_id": str(node_id), **detail})
    except ImportError:
        pass


class MengBaoImageReverse:
    @classmethod
    def INPUT_TYPES(cls):
        return {"required": {"image": ("IMAGE",), "request_id": ("STRING", {"default": ""})}, "hidden": {"unique_id": "UNIQUE_ID"}}

    RETURN_TYPES = ("MENGBAO_REPLICA_ANALYSIS", "STRING")
    RETURN_NAMES = ("analysis", "reverse_prompt")
    FUNCTION = "analyze"
    OUTPUT_NODE = True
    CATEGORY = "萌宝AI/电商"
    DESCRIPTION = "识别图片构图与可替换元素；使用 GEM 3.7 flash，连接失败切换 GEM 3.8 flash。"
    SEARCH_ALIASES = ["Meng", "MengBao", "MengBaoAI", "MengBao AI", "MengBaoImageReverse", "萌宝", "萌宝AI", "萌宝AI·图片反推", "图片反推", "图片复刻", "Image Reverse", "Image Replication", "MengBao AI · Image Reverse"]

    def analyze(self, image, request_id, unique_id=None):
        if not request_id:
            raise ValueError("Click Analyze Image in the Image Reverse node first")
        identity(request_id)
        if not isinstance(image, torch.Tensor) or image.dim() != 4 or image.shape[0] != 1:
            raise ValueError("Image Reverse requires exactly one IMAGE, not an image batch")
        source = REPLICA_STORE.put_asset(image_tensor_to_png_bytes(image), "source.png")
        fingerprint = f"analysis:{source['id']}:{PROMPT_VERSION}"
        record = REPLICA_STORE.begin(request_id, fingerprint)
        if record is None:
            try:
                result = run_vision(ANALYSIS_PROMPT, [REPLICA_STORE.asset_bytes(source["id"])], parse_analysis,
                                    on_stage=lambda event: status(unique_id, {**event, "request_id": request_id}), cancelled=interrupted)
                record = {"id": request_id, "source": source, "analysis": result["data"], "model": result["model"], "attempts": result["attempts"]}
                REPLICA_STORE.complete(request_id, record)
            except BaseException as exc:
                REPLICA_STORE.fail(request_id, str(exc) or "Vision request interrupted")
                raise
        prompt = "\n\n".join([record["analysis"]["composition"], *[
            f"{element['name']}: {element['description']}\n{element['text']}\n{element['textStyle']}" for element in record["analysis"]["elements"]]])
        status(unique_id, {"stage": "completed", "model": record["model"]})
        return {"ui": {"mengbao_replica_analysis": [record]}, "result": (record, prompt)}


class MengBaoImageReplicaSettings:
    @classmethod
    def INPUT_TYPES(cls):
        return {"required": {"analysis": ("MENGBAO_REPLICA_ANALYSIS",), "draft_json": ("STRING", {"default": ""}), "draft_id": ("STRING", {"default": ""})}}

    RETURN_TYPES = ("STRING", "MENGBAO_REPLICA_SETTINGS")
    RETURN_NAMES = ("prompt", "replica_settings")
    FUNCTION = "build"
    CATEGORY = "萌宝AI/电商"
    DESCRIPTION = "可视化校正图片元素、修改文字、替换素材，确认后连接萌宝AI·图像生成。"
    SEARCH_ALIASES = ["Meng", "MengBao", "MengBaoAI", "MengBao AI", "MengBaoImageReplicaSettings", "萌宝", "萌宝AI", "萌宝AI·图片复刻设置", "复刻设置", "图片复刻", "元素替换", "Replica Settings", "MengBao AI · Replica Settings"]

    def build(self, analysis, draft_json, draft_id):
        settings = compile_settings(analysis, draft_json, REPLICA_STORE)
        REPLICA_STORE.save_draft(draft_id, json.loads(draft_json))
        return settings["prompt"], settings


class MengBaoReplicaAudit:
    @classmethod
    def INPUT_TYPES(cls):
        return {"required": {"images": ("IMAGE",), "replica_settings": ("MENGBAO_REPLICA_SETTINGS",)}, "hidden": {"unique_id": "UNIQUE_ID"}}

    RETURN_TYPES = ("IMAGE", "STRING")
    RETURN_NAMES = ("images", "report")
    FUNCTION = "audit"
    OUTPUT_NODE = True
    CATEGORY = "萌宝AI/电商"
    DESCRIPTION = "检查复刻成图缺字、改字和旧文案残留；检查失败仍保留生成图片。"
    SEARCH_ALIASES = ["Meng", "MengBao", "MengBaoAI", "MengBao AI", "MengBaoReplicaAudit", "萌宝", "萌宝AI", "萌宝AI·复刻检查", "复刻检查", "文字检查", "Replica Audit", "MengBao AI · Replica Audit"]

    def audit(self, images, replica_settings, unique_id=None):
        reports = []
        for index, image in enumerate(images):
            if interrupted():
                reports.append({"image": index + 1, "error": "Vision request cancelled"})
                break
            try:
                result = run_vision(audit_prompt(replica_settings), [image_tensor_to_png_bytes(image)], parse_audit, max_tokens=3000,
                                    on_stage=lambda event: status(unique_id, {**event, "image": index + 1}), cancelled=interrupted)
                reports.append({"image": index + 1, "model": result["model"], "attempts": result["attempts"], **result["data"]})
            except Exception as exc:
                reports.append({"image": index + 1, "error": str(exc)})
        report = json.dumps(reports, ensure_ascii=False, indent=2)
        return {"ui": {"mengbao_replica_report": [report]}, "result": (images, report)}
